core/base/src/TDirectory.cxx

```diff
@@ -381,16 +412,28 @@ TObject *TDirectory::CloneObject(const TObject *obj, Bool_t autoadd /* = kTRUE *
    return newobj;
 }
 
+////////////////////////////////////////////////////////////////////////////////
+/// Return the (address of) a shared pointer to the struct holding the
+/// actual thread local gDirectory pointer and the atomic_flag for its lock.
+std::shared_ptr<TDirectory::TGDirectory> &TDirectory::GetSharedLocalCurrentDirectory()
+{
+   using shared_ptr_type = std::shared_ptr<TDirectory::TGDirectory>;
+
+   // Note in previous implementation every time gDirectory was lookup in
+   // a thread, if it was set to nullptr it would be reset to gROOT.  This
+   // was unexpected and this routine is not re-introducing this issue.
```

Comment:
Great. I don't see that. `GetSharedLocalCurrentDirectory()` is `gDirectory`, and it returns `gROOTLocal`. What am I missing?

Reply:
The thread local initialization is executed only once, so this routine only set the initial value of a given thread's `gDirectory` to `gROOTLocal` but otherwise (all but after the first invocation in a thread) leave it unchanged.

The previous version of the code essential had:
```
thread_local TDirectory *currentDirectory = nullptr;
if (currentDirectory == nullptr)
   currentDirectory = ROOT::Internal::gROOTLocal;
```